vowpalwabbit/csoaa.cc

```diff
@@ -129,7 +145,69 @@ void predict_or_learn(csoaa& c, single_learner& base, example& ec)
   ec.pred.multiclass = prediction;
 }
 
-void finish_example(vw& all, csoaa&, example& ec) { COST_SENSITIVE::finish_example(all, ec); }
+template <bool probabilities>
+void finish_example(vw& all, csoaa& c, example& ec)
+{
+  if (probabilities)
+  {
+    // === Compute multiclass_log_loss
+    // TODO:
+    // What to do if the correct label is unknown, i.e. (uint32_t)-1?
+    //   Suggestion: increase all.sd->weighted_unlabeled_examples???,
+    //               but not sd.example_number, so the average loss is not influenced.
+    // What to do if the correct_class_prob==0?
+    //   Suggestion: have some maximal multiclass_log_loss limit, e.g. 999.
+    float multiclass_log_loss = 999;  // -log(0) = plus infinity
+    float correct_class_prob = 0;
+    if (ec.l.multi.label <= c.num_classes)  // prevent segmentation fault if labeĺ==(uint32_t)-1
+      correct_class_prob = ec.pred.scalars[ec.l.multi.label - 1];
+    if (correct_class_prob > 0) multiclass_log_loss = -std::log(correct_class_prob) * ec.weight;
+    if (ec.test_only)
+      all.sd->holdout_multiclass_log_loss += multiclass_log_loss;
+    else
+      all.sd->multiclass_log_loss += multiclass_log_loss;
+
+    // === Compute `prediction` and zero_one_loss
+    // We have already computed `prediction` in predict_or_learn,
+    // but we cannot store it in ec.pred union because we store ec.pred.probs there.
+    uint32_t prediction = 0;
+    for (uint32_t i = 1; i < c.num_classes; i++)
+      if (ec.pred.scalars[i] > ec.pred.scalars[prediction]) prediction = i;
+    prediction++;  // prediction is 1-based index (not 0-based)
+    float zero_one_loss = 0;
+    if (ec.l.multi.label != prediction) zero_one_loss = ec.weight;
+
+    // === Print probabilities for all classes
+    std::ostringstream outputStringStream;
+    for (uint32_t i = 0; i < c.num_classes; i++)
+    {
+      if (i > 0) outputStringStream << ' ';
+      if (all.sd->ldict) { outputStringStream << all.sd->ldict->get(i + 1); }
```

Comment:
I don't think ldict is used in csoaa

Reply:
This is used for the named case, ie in `test/pred-sets/ref/csoaa_named_probabilities.predict` it will output the names of the labels instead of the indices